source/extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include <iostream>
+#include <memory>
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+
+#include "quiche/quic/core/quic_buffer_allocator.h"
+
+namespace quic {
+
+// Implements the interface required by
+// https://quiche.googlesource.com/quiche/+/refs/heads/master/quic/platform/api/quic_mem_slice.h
+// Note the Quic plugin strategy relies on link-time binding rather than inheritance.
+class QuicMemSliceImpl {
+public:
+  // Constructs an empty QuicMemSliceImpl.
+  QuicMemSliceImpl() = default;
+
+  // Constructs a QuicMemSliceImpl by let |allocator| allocate a data buffer of
+  // |length|.
+  QuicMemSliceImpl(QuicBufferAllocator* allocator, size_t length) {
+    size_t total_length = sizeof(Envoy::Buffer::BufferFragmentImpl) + length;
+    char* mem = allocator->New(total_length);
+    auto fragment = new (mem + length) Envoy::Buffer::BufferFragmentImpl(
```

Comment:
The fragment may not be properly aligned.  How about put fragment in the beginning?

// In a anonymous namespace.
struct BufferFragmentBundle {
  // The alignas() is not necessary, but it makes |buffer| to start at a max aligned address, which is nice.
  struct {
    alignas(std::max_align_t) Envoy::Buffer::BufferFragmentImpl fragment;
  } fragment_with_padding;
  char buffer[0];
};

QuicMemSliceImpl(QuicBufferAllocator* allocator, size_t length) {
  BufferFragmentBundle* bundle = allocator->New(sizeof(BufferFragmentBundle) + length);
  auto fragment = new (&bundle->fragment_with_padding.fragment) Envoy::Buffer::BufferFragmentImpl(
      bundle->buffer, length, ...);
};